src/components/sidecar/use-active-section.ts

```diff
@@ -0,0 +1,100 @@
+import { useEffect, useRef, useState } from 'react'
+import { SidecarHeading } from './types'
+
+/**
+ * This code comes from one of Bryce's (@BRKalow) @hashicorp/react-components PRs:
+ * https://github.com/hashicorp/react-components/pull/325
+ */
+export function useActiveSection(
+  headings: SidecarHeading[],
+  isEnabled = true
+): string {
+  const [activeSection, setActiveSection] = useState<SidecarHeading['slug']>()
+  const previousY = useRef<number>()
+
+  useEffect(() => {
+    if (!isEnabled) {
+      return
+    }
+
+    const findMatchingSectionIndex = (slug: string) => {
+      return headings.findIndex((section) => section.slug === slug)
+    }
+
+    const observer = new IntersectionObserver(
+      (entries) => {
+        let currentY: number
+        let scrollTrend: 'down' | 'up'
+        const visibleHeadings = []
+
+        entries.forEach((entry) => {
+          currentY = window.scrollY
+          scrollTrend = previousY.current < currentY ? 'down' : 'up'
+
+          if (entry.isIntersecting) {
+            visibleHeadings.push(entry.target.id)
+          }
+        })
+
+        const isSingleEntryLeaving =
+          entries.length === 1 && !entries[0].isIntersecting
+        const singleEntryLeavingIndex = isSingleEntryLeaving
+          ? findMatchingSectionIndex(entries[0].target.id)
+          : -1
+
+        // Activate only the bottom-most visible section heading
+        if (visibleHeadings.length === 1) {
+          setActiveSection(visibleHeadings[0])
+        } else if (visibleHeadings.length > 1) {
+          setActiveSection(visibleHeadings[visibleHeadings.length - 1])
+        }
+
+        if (previousY.current) {
+          // If we detect that we're scrolling up, and there are no visible headers,
+          // optimistically set the previous header as visible to make the active section match the visible content
+          if (visibleHeadings.length === 0 && scrollTrend === 'up') {
+            setActiveSection((current) => {
+              const curActiveIndex = findMatchingSectionIndex(current)
+
+              // Handle an ege case where we get an intersection event for a heading further down the page
+              // leaving intersection, otherwise this would cause the active heading to incorrectly get bumped up
+              if (
+                isSingleEntryLeaving &&
+                singleEntryLeavingIndex > curActiveIndex
+              ) {
+                return current
+              }
+
+              const newIndex = curActiveIndex - 1
+
+              if (newIndex < 0) {
+                return current
+              }
+
+              return headings[newIndex].slug
+            })
+          }
+        }
+
+        if (currentY) {
+          previousY.current = currentY
+        }
+      },
+      { rootMargin: '0% 0% -75% 0%' }
```

Comment:
I think we need to tweak this number some more, here's one example where the next heading is highlighted after clicking on a link: https://dev-portal-git-ambimprove-sidecar-and-responsiveness-hashicorp.vercel.app/waypoint/docs#getting-started

Reply:
Noticing Getting started doesn't get hit here. Not sure if that'd be related/fixed in a root margin change 

https://user-images.githubusercontent.com/825855/146246232-d8ebbd87-85a7-497a-a686-c7e3a52cc007.mp4

Reply:
Yes, that should be fixed with the root margin change! It's related at the very least